### ISSUE TITLE
Versioning

### ISSUE THREAD

**Issue by j-c-cook** [MassimoCimmino/pygfunction#205]
Currently, the version in `setup` is not changed until a new release. This means that if the `main` branch is installed, then pip will inform the user that it is a previously released version, rather than the development version of the next release. Directly following a release, should we edit the version to be `vx.x.x-dev`?

- Last commit message prior to a release, the `version` gets changed (what we do now)
- The first commit after a release, we mark the `version` as `dev` for the next release (what I am asking if we should do)

**Comment by MassimoCimmino** [comment]
This is a good point.

There is also the option of incrementing the dev version at each merge on the `master`. There should be tools to do this automatically. It is worth looking into.

**Comment by MassimoCimmino** [comment]
Tools like [python-semantic-release](https://python-semantic-release.readthedocs.io/en/latest/) exist for automated versioning but seem to much for pygfunction (there is simply not enough development).

Incrementing the main branch with `.dev0` is an option but may create problems if a hotfix is needed to the latest minor version.

A solution would be to introduce a Gitflow-like workflow to manage branches. `master` is only for for releases. `feature` branches are merged to an additional `develop` branch. This way `hotfix` and maintenance branches can be cleanly created abd merged into `master`.

We can implement this following the upcoming release of `v2.2`.

**Comment by j-c-cook** [comment]
I agree with your proposed solution. I like the idea of being able to merge hot fixes into the release branch (contents being the most recently released version) for quick bug releases. 

Should `master` be renamed `release`?

**Comment by j-c-cook** [comment]
Here's a post describing the gitflow process: https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow

Here's a post that argues against it: https://georgestocker.com/2020/03/04/please-stop-recommending-git-flow/

**Comment by MassimoCimmino** [comment]
The second post makes a good argument against Gitflow.

Alternatives:

- [Trunk Based Development](https://trunkbaseddevelopment.com/)
- [The cactus model](https://barro.github.io/2016/02/a-succesful-git-branching-model-considered-harmful/)
- [Three-Flow](https://www.nomachetejuggling.com/2017/04/09/a-different-branching-strategy/)

All three are variations of the same idea.

**Comment by MassimoCimmino** [comment]
Trunk Based Development seems simple enough and close to the current workflow. We are already using "short-lived" feature branches and releasing from a tag on the trunk (`master`).

When needed, release branches can be created retroactively from the tag for any needed bug fixes. These would be cherrypicked from the trunk (good practice is to avoid working on release branches).

It does not look like there is much to be changed here with regards to the workflow. I can add a notice/link in `CONTRIBUTING.md` to mention trunk based development.

For versioning, here I propose the first commit after a release be to increment the version to `X.(Y+1).0.dev0` to start the next release cycle. This would be done through a PR but does not need to have an associated issue. Issues that introduce major changes could directly increment to `(X+1).0.0.dev0`.
